scripts/apps/authoring/authoring/directives/AuthoringDirective.ts

```diff
@@ -846,18 +846,40 @@ export function AuthoringDirective(
                 $scope.autosave(item);
             };
 
+            Object.values(extensions).forEach((extension) => {
+                if (extension.activationResult?.contributions?.authoring != null) {
+                    $scope.onUpdateFromExtension = extension.activationResult.contributions.authoring.onUpdate;
+                }
+            });
+
+            function performOnUpdateFromExtensionAndAutosave(item, origItem, timeout) {
+                $scope.onUpdateFromExtension($scope.origItem._autosave ?? $scope.origItem, $scope.item)
+                    .then((article) => {
+                        angular.extend($scope.item, article);
+                        var autosavedItem = authoring.autosave(item, origItem, timeout);
+
+                        authoringWorkspace.addAutosave();
+                        initMedia();
+                        updateSchema();
+                        return autosavedItem;
+                    });
+            }
+
             $scope.autosave = function(item, timeout) {
                 $scope.dirty = true;
                 angular.extend($scope.item, item); // make sure all changes are available
                 return coreApplyMiddleware(onChangeMiddleware, {item: $scope.item, original: $scope.origItem}, 'item')
                     .then(() => {
-                        var autosavedItem = authoring.autosave($scope.item, $scope.origItem, timeout);
-
-                        authoringWorkspace.addAutosave();
+                        if ($scope.onUpdateFromExtension != null) {
+                            performOnUpdateFromExtensionAndAutosave($scope.item, $scope.origItem, timeout);
+                        } else {
+                            var autosavedItem = authoring.autosave($scope.item, $scope.origItem, timeout);
 
-                        initMedia();
-                        updateSchema();
-                        return autosavedItem;
+                            authoringWorkspace.addAutosave();
+                            initMedia();
+                            updateSchema();
+                            return autosavedItem;
```

Comment:
You are duplicating this code in `performOnUpdateFromExtensionAndAutosave`. Actually, you don't need neither `performOnUpdateFromExtensionAndAutosave` as a separate function nor `onUpdateFromExtension` variable in the scope.

You can do it like this

```
const onUpdateFromExtensions = Object.values(extensions).map(...);

(
    onUpdateFromExtensions.length < 1
    ? Promise.resolve(item)
    : onUpdateFromExtensions.reduce(...).then((nextItem) => {
        angular.extend($scope.item, nextItem);
    })
).then(() => {
    var autosavedItem = authoring.autosave($scope.item, $scope.origItem, timeout);

    authoringWorkspace.addAutosave();
    initMedia();
    updateSchema();
    return autosavedItem;
});
```